internal/streamingnode/server/flusher/flusherimpl/flusher_impl.go

```diff
@@ -0,0 +1,203 @@
+		return err
+	}
+	for _, vchannel := range resp.GetVchannels() {
+		f.tasks.Insert(vchannel, wal)
+	}
+	return nil
+}
+
+func (f *flusherImpl) UnregisterPChannel(pchannel string) {
+	f.scanners.Range(func(vchannel string, scanner wal.Scanner) bool {
+		if funcutil.ToPhysicalChannel(vchannel) != pchannel {
+			return true
+		}
+		f.UnregisterVChannel(vchannel)
+		return true
+	})
+}
+
+func (f *flusherImpl) RegisterVChannel(vchannel string, wal wal.WAL) {
+	f.tasks.Insert(vchannel, wal)
+}
+
+func (f *flusherImpl) UnregisterVChannel(vchannel string) {
+	if scanner, ok := f.scanners.GetAndRemove(vchannel); ok {
+		err := scanner.Close()
+		if err != nil {
+			log.Warn("scanner error", zap.String("vchannel", vchannel), zap.Error(err))
+		}
+	}
+	f.fgMgr.RemoveFlowgraph(vchannel)
+	f.wbMgr.RemoveChannel(vchannel)
+}
+
+func (f *flusherImpl) Start() {
+	f.wbMgr.Start()
+	go f.cpUpdater.Start()
+	go func() {
+		ticker := time.NewTicker(tickDuration)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-f.stopChan:
+				log.Info("flusher stopped")
+				return
+			case <-ticker.C:
+				f.tasks.Range(func(vchannel string, wal wal.WAL) bool {
+					err := f.buildPipeline(vchannel, wal)
+					if err != nil {
+						log.Warn("build pipeline failed", zap.String("vchannel", vchannel), zap.Error(err))
+						return true
+					}
+					log.Info("build pipeline done", zap.String("vchannel", vchannel))
+					return true
+				})
+			}
+		}
+	}()
+}
+
+func (f *flusherImpl) Stop() {
+	f.stopOnce.Do(func() {
+		close(f.stopChan)
+		f.scanners.Range(func(vchannel string, scanner wal.Scanner) bool {
+			err := scanner.Close()
+			if err != nil {
+				log.Warn("scanner error", zap.String("vchannel", vchannel), zap.Error(err))
+			}
+			return true
+		})
+		f.fgMgr.ClearFlowgraphs()
+		f.wbMgr.Stop()
+		f.cpUpdater.Close()
+	})
+}
+
+func (f *flusherImpl) buildPipeline(vchannel string, w wal.WAL) error {
+	if f.fgMgr.HasFlowgraph(vchannel) {
+		return nil
+	}
+
+	// Get recovery info from datacoord.
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*30)
+	defer cancel()
+	resp, err := resource.Resource().DataCoordClient().
+		GetChannelRecoveryInfo(ctx, &datapb.GetChannelRecoveryInfoRequest{Vchannel: vchannel})
+	if err != nil {
+		return err
+	}
+
+	// Convert common.MessageID to message.messageID.
+	mqWrapperID, err := adaptor.DeserializeToMQWrapperID(resp.GetInfo().GetSeekPosition().GetMsgID(), w.WALName())
+	if err != nil {
+		return err
+	}
+	messageID := adaptor.MustGetMessageIDFromMQWrapperID(mqWrapperID)
+
+	// Create scanner.
+	policy := options.DeliverPolicyStartFrom(messageID)
+	filter := func(msg message.ImmutableMessage) bool { return msg.VChannel() == vchannel }
+	handler := adaptor2.NewMsgPackAdaptorHandler()
+	ro := wal.ReadOption{
+		DeliverPolicy:  policy,
+		MessageFilter:  filter,
+		MesasgeHandler: handler,
+	}
+	scanner, err := w.Read(ctx, ro)
+	if err != nil {
+		return err
+	}
+	f.scanners.Insert(vchannel, scanner)
```

Comment:
move it to line 201 to prevent a potential leak in f.scanner if return error at line 198